File: python/other/bst/flatten-linked-list.py
class Solution:
    def flatten(self, root):
        cur = root
        while cur:
            if cur.left:
                last = cur.left
                while last.right:
                    last = last.right
                last.right = cur.right
                cur.right = cur.left
                cur.left = None
            cur = cur.right

class Solution2:
    def __init__(self):
        self.prev = None
    
    def flatten(self, root):
        if not root:
            return None
        self.flatten(root.right)
        self.flatten(root.left)
    
        root.right = self.prev
        root.left = None
        self.prev = root

File: go/other/bst/flatten-linked-list.go
package bst

func Flatten(root *TreeNode) *TreeNode {
	cur := root
	for cur != nil {
		if cur.Left != nil {
			last := cur.Left
			for last.Right != nil {
				last = last.Right
			}
			last.Right = cur.Right
			cur.Right = cur.Left
			cur.Left = nil
		}
		cur = cur.Right
	}
	return nil
}
